plugins/out_s3/s3.c

```diff
@@ -1734,22 +1818,18 @@ static void s3_upload_queue(struct flb_config *config, void *out_context)
             s3_store_file_lock(upload_contents->upload_file);
             ctx->upload_queue_success = FLB_FALSE;
 
-            /* If retry limit was reached, discard file and remove file from queue */
-            upload_contents->retry_counter++;
-            if (upload_contents->retry_counter >= MAX_UPLOAD_ERRORS) {
-                flb_plg_warn(ctx->ins, "Chunk file failed to send %d times, will not "
-                             "retry", upload_contents->retry_counter);
-                s3_store_file_inactive(ctx, upload_contents->upload_file);
-                multipart_upload_destroy(upload_contents->m_upload_file);
+            /* If retry limit was reached, remove file from queue */
+            if (upload_contents->upload_file->failures > ctx->ins->retry_limit) {
                 remove_from_queue(upload_contents);
                 continue;
             }
 
             /* Retry in N seconds */
-            upload_contents->upload_time = now + 2 * upload_contents->retry_counter;
-            ctx->retry_time += 2 * upload_contents->retry_counter;
+            upload_contents->upload_time = now + 2 * upload_contents->upload_file->failures;
+            ctx->retry_time += 2 * upload_contents->upload_file->failures;
```

Comment:
This is not exponentual backoff, exponentual backoff would be 
```
if (ctx->retry_time == 0) {
   retry_time = 1
}
else {
   ctx->retry_time += 2 * ctx->retry_time;
}
upload_contents->upload_time = now + ctx->retry_time;
```

Reply:
It's totally fine. Just a comment in case we didn't realize this.

increasing by a set multiple is prone to getting a spam of requests every 2 seconds to constantly fail repeatedly due to piling up requests

Reply:
Hmm wait- we should remove the comment, this code does nothing I think? `ctx->retry_time` and ctx->upload_time` don't seem to be used AFAICT.

Reply:
Are you sure that we don't need to set retry_time and upload_timeout? Those variables are still being used to trigger retries:
https://github.com/fluent/fluent-bit/blob/760956f50cdc7eef3047df20d2299202a8c68594/plugins/out_s3/s3.c#L1020
https://github.com/fluent/fluent-bit/blob/760956f50cdc7eef3047df20d2299202a8c68594/plugins/out_s3/s3.c#L1154
https://github.com/fluent/fluent-bit/blob/760956f50cdc7eef3047df20d2299202a8c68594/plugins/out_s3/s3.c#L1794
https://github.com/fluent/fluent-bit/blob/760956f50cdc7eef3047df20d2299202a8c68594/plugins/out_s3/s3.c#L1842